### PR TITLE
[do-not-review] [cudagraph] Warm up two optimizer steps before capture

### PR DESCRIPTION
**Pull request by xmfan**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* (to be filled)

## Why

One eager forward-backward invocation is not enough to establish steady-state training storage. Optimizers may allocate state lazily during their first `step()`, and gradient accumulation may invoke forward-backward several times before one optimizer step.

Capturing before those allocations stabilize can leave persistent allocations outside the graph pool or record a non-steady-state memory layout.

## Capture timing

```text
Optimizer step 1
    +--> gradient-accumulation forward/backward calls
    +--> scheduled SDC replay calls
    +--> lazy optimizer-state allocation

Optimizer step 2
    +--> gradient-accumulation forward/backward calls
    +--> scheduled SDC replay calls
    +--> steady-state allocator behavior

Optimizer step 3
    +--> capture the next forward/backward invocation
    +--> replay thereafter
```

The warmup count is measured in wrapped forward-backward invocations:

```text
2 * gradient_accumulation_steps
+ checked_warmup_steps * sdc_num_replays
```

## Change

Add a validated `num_warmup_iterations` option to `CUDAGraphWrapper` and its convenience wrapper. The Trainer computes the number of invocations performed during the first two optimizer steps, including additional executions made by scheduled SDC replay.

Capture therefore begins only after optimizer state, gradients, and allocator behavior have reached their steady-state layout.

## Testing

The tests cover explicit and invalid warmup counts, gradient accumulation, finite and continuous SDC schedules, and forwarding the configured count into the graph wrapper.

- `python -m pytest -q tests/unit_tests/cpu/test_cudagraph.py tests/unit_tests/cpu/test_trainer.py`
- `python -m tests.integration_tests.run_tests <output> --test_suite features --test_name sdc_replay_cudagraph --execution_mode real_pg --ngpu 1 --no-parallel` (capture begins on step 3)
- Full stack: four-GPU `pp_looped_1f1b_cudagraph` integration, eager steps 1-2, capture on step 3, replay through step 10

stack-info: PR: https://github.com/pytorch/torchtitan/pull/4433, branch: xmfan/stack/50

### CODE REVIEW

**Comment by xmfan** [comment]
Closing duplicate PR created by an incorrect ghstack resubmission; the original stack remains in #4429-#4436.
